keyring.go

```diff
@@ -193,23 +193,29 @@ func (k KeyRing) VerifyJSONs(ctx context.Context, requests []VerifyJSONRequest)
 		return nil, err
 	}
 
-	// TODO: we should distinguish here between expired keys, and those we don't have.
-	// If the key has expired, it's no use re-requesting it.
-	now := AsTimestamp(time.Now())
+	// Default to not hitting federation.
+	doFederationHit := false
+
+	// TODO: we should distinguish here between expired keys, and those we
+	// don't have. If the key has expired, it's no use re-requesting it.
 	keysFetched := map[PublicKeyLookupRequest]PublicKeyLookupResult{}
+	now := AsTimestamp(time.Now())
 	for req, res := range keysFromDatabase {
+		if res.ExpiredTS != PublicKeyNotExpired {
+			// Don't bother re-requesting keys that are marked as expired.
+			delete(keyRequests, req)
+		}
 		if now > res.ValidUntilTS && res.ExpiredTS == PublicKeyNotExpired {
-			// We have passed the key validity period so we should re-request from
-			// the remote server (or the perspectives).
-			continue
+			// We're past the validity for this key so we should really
+			// hit the fetchers regardless to renew it.
+			doFederationHit = true
```

Comment:
If this is `true` then why bother doing `checkUsingKeys` on `:216`? It's just wasted effort as it will never then return early.

Reply:
I don't see how setting this to true helps - if the key is bad the check will fail and THEN we'll hit federation?

Reply:
I've shuffled this around a bit now - we'll only try the first `checkUsingKeys` if we think all of our key requests are satisfied. Then if we *still* need to hit federation, we will do.

Reply:
I don't see what `doFederationHit` buys us in this part of the code. It would be just as good to say `doFederationHit := len(keysFetched) < len(keyRequests)` which is also clearer IMO.